service/src/main/java/org/cbioportal/service/impl/GenesetHierarchyServiceImpl.java

```diff
@@ -0,0 +1,385 @@
+			if (genesetDataMap.get(geneset.getGenesetId()) != null) {
+				//there is data, so add this one:
+				result.add(geneset);
+			}
+		}
+		return result;
+	}
+
+	/**
+	 * This will set the representativeScore attribute for each gene set, based on the data (the gene set 
+	 * scores per sample).
+	 * @param sampleIds 
+	 * 
+	 * @param genesets: list of gene sets for which to calculate and record the representativeScore
+	 * @param genesetDataMap: the set of GSVA(like) scores per sample for each gene set 
+	 * @param percentile: (optional) which percentile to use when determining the representativeScore. If not 
+	 *            set, max of absolute score is returned.
+	 * @throws GeneticProfileNotFoundException 
+	 */
+	private void fillRepresentativeScoresAndPvalues(List<Geneset> genesets, 
+			Map<String, List<GenesetGeneticData>> genesetScoresMap, Map<String, List<GenesetGeneticData>> genesetPvaluesMap, 
+			Integer percentile) throws GeneticProfileNotFoundException {
+		
+		genesets.stream().forEach(g -> calculateAndSetRepresentativeScoreAndPvalue(g, genesetScoresMap, genesetPvaluesMap, percentile));
+	}
+
+	private void calculateAndSetRepresentativeScoreAndPvalue(Geneset geneset, 
+			Map<String, List<GenesetGeneticData>> genesetScoresMap, 
+			Map<String, List<GenesetGeneticData>> genesetPvaluesMap, 
+			Integer percentile) {
+		
+		List<GenesetGeneticData> genesetScoreData = genesetScoresMap.get(geneset.getGenesetId());
+		List<GenesetGeneticData> genesetPvalueData = genesetPvaluesMap.get(geneset.getGenesetId());
+		
+		//return the maximum absolute value found:
+		List<ScoreAndPvalue> positiveScoresAndPvalues = new ArrayList<ScoreAndPvalue>();
+		List<ScoreAndPvalue> negativeScoresAndPvalues = new ArrayList<ScoreAndPvalue>();
+		
+		double max = 0;
+		double pvalueOfMax = 1;
+		for (int i = 0; i < genesetScoreData.size(); i++) {
+			String scoreString = genesetScoreData.get(i).getValue();
+			String pvalueString = genesetPvalueData.get(i).getValue();
+			
+			if (!NumberUtils.isNumber(scoreString)) 
+    			continue;
+    		
+			double score = Double.parseDouble(scoreString);
+			double pvalue = 1.0;
+			if (NumberUtils.isNumber(pvalueString))
+				pvalue = Double.parseDouble(pvalueString);
+			if (score >= 0) {
+				positiveScoresAndPvalues.add(new ScoreAndPvalue(score, pvalue));
+			} else {
+				negativeScoresAndPvalues.add(new ScoreAndPvalue(score, pvalue));
+			}
+			
+			//keep track of max, in case percentile is null
+    		if (Math.abs(score) > Math.abs(max)) {
+    			max = score; //here no abs, since we want to get the raw score (could be negative)
+    			pvalueOfMax = pvalue;
+    		}
+		}
+		
+		if (percentile == null) {
+			geneset.setRepresentativeScore(max);
+			geneset.setRepresentativePvalue(pvalueOfMax);
+		} else {
+			//sort scores:
+			Collections.sort(positiveScoresAndPvalues, ScoreAndPvalue.comparator);
+			Collections.sort(negativeScoresAndPvalues, ScoreAndPvalue.comparatorDesc);
```

Comment:
Can you use lambda expression here using `positiveScoresAndPvalues.sort` method, instead of Comparator?